packages/react-ui-core/src/Gmap/FreeDrawLayer.js

```diff
@@ -0,0 +1,230 @@
+        this.formatLongLatToGmapsCoordinates(
+          this.props.shapes
+        )[0]
+      )
+    }
+  }
+
+  getPolygonCoords(poly) {
+    const len = poly.getPath().getLength()
+    const shape = []
+
+    for (let i = 0; i < len; i += 1) {
+      shape.push(
+        poly.getPath()
+          .getAt(i)
+          .toUrlValue(5)
+          .split(',')
+          .map(Number)
+          .reverse()
+      )
+    }
+    return shape
+  }
+
+  enableMapControls() {
+    const { map, mapControls } = this.props
+    map.setOptions({
+      draggable: true,
+      zoomControl: true,
+      scrollwheel: true,
+      disableDoubleClickZoom: true,
+      ...mapControls,
+    })
+    this.setState({
+      drawing: false,
+    })
+  }
+
+  disableMapControls() {
+    const { map } = this.props
+    map.setOptions({
+      draggable: false,
+      zoomControl: false,
+      scrollwheel: false,
+      disableDoubleClickZoom: false,
+    })
+    this.setState({
+      drawing: true,
+    })
+  }
+
+  @autobind
+  enableMapDraw() {
+    const { map } = this.props
+    this.disableMapControls()
+    setupEvents(map, EVENTS, {
+      onMouseDown: () => {
+        this.drawFreeHand()
+      },
+    }, true)
+  }
+
+  drawFreeHand() {
+    const { map, onMapDrawStart, multipleShapes, dataStyle } = this.props
+    const poly = new window.google.maps.Polyline({
+      map,
+      clickable: false,
+      ...dataStyle.polyline,
+    })
+    const drawPoly = new window.google.maps.Polygon({
+      map,
+      path: poly.getPath(),
+      clickable: false,
+      ...dataStyle.polylineFill,
+    })
+
+    if (onMapDrawStart) {
+      onMapDrawStart()
+    }
+
+    if (!multipleShapes) {
+      this.clearAllShapes()
+    }
+
+    const listener = setupEvents(map, EVENTS, {
+      onMouseMove: e => {
+        poly.getPath().push(e.latLng)
+      },
+      onTouchStart: e => {
+        poly.getPath().push(e.latLng)
+      },
+    }, false)
+
+    setupEvents(map, EVENTS, {
+      onMouseUp: () => {
+        removeEvent(listener.onMouseMove)
+        poly.setMap(null)
+        drawPoly.setMap(null)
+        this.addShape(poly.getPath())
+        this.enableMapControls()
+      },
+      onTouchEnd: () => {
+        removeEvent(listener.onMouseMove)
+        poly.setMap(null)
+        drawPoly.setMap(null)
+        this.addShape(poly.getPath())
+        this.enableMapControls()
+      },
+    }, true)
+  }
+
+  addShapeToGMaps(path) {
+    const { map, dataStyle } = this.props
+    return new window.google.maps.Polygon({
+      map,
+      path,
+      ...dataStyle.polygon,
+    })
+  }
+
+  addShape(path) {
+    const { onMapDrawEnd } = this.props
+    const { polygons } = this.state
+
+    const rawPolygon = this.addShapeToGMaps(path)
+
+    this.setState({
+      polygons: [...polygons, rawPolygon],
+    }, () => {
+      if (onMapDrawEnd) {
+        onMapDrawEnd(this.state.polygons.reduce((acc, polygon, i) => {
+          acc[i] = this.getPolygonCoords(polygon)
```

Comment:
We shouldn't be modifying the prop `acc`, we should be avoiding mutation. Can we find a different way to write this?

Reply:
The `acc` param is the accumulated value from the reduce function - so it's not the prop though...

Can we store the path lat long as such in the state => this.state.polygon = [[lng, lat], [lng, lat], ...]? That way if Polygon objects need to be created the FreeDrawLayer component or the GMap component will then use those path values to create distinct Polygon Google Map objects when needed. Storing the actual external api in state then forces us to have difficult means of using the data that we need (just the path values). We wouldn't need to reduce them, we'd just map over them and create Polygon objects when need be. I could be wrong here since I'm not as familiar with the Polygon Maps api so let me know if it's needed to store the entire object in React's state.